src/shims/sync.rs

```diff
@@ -0,0 +1,438 @@
+    // bytes 4-7: count of how many times this mutex has been locked, as a u32
+    // bytes 12-15: mutex kind, as an i32
+    // (the kind should be at this offset for compatibility with the static
+    // initializer macro)
+
+    fn pthread_mutex_init(
+        &mut self,
+        mutex_op: OpTy<'tcx, Tag>,
+        attr_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        check_ptr_target_min_size(this, mutex_op, 16)?;
+        check_ptr_target_min_size(this, attr_op, 4)?;
+
+        let mutex = this.read_scalar(mutex_op)?.not_undef()?;
+        if this.is_null(mutex)? {
+            return this.eval_libc_i32("EINVAL");
+        }
+        let mutex_place = this.deref_operand(mutex_op)?;
+
+        let i32_layout = this.layout_of(this.tcx.types.i32)?;
+
+        let attr = this.read_scalar(attr_op)?.not_undef()?;
+        let kind = if this.is_null(attr)? {
+            this.eval_libc("PTHREAD_MUTEX_DEFAULT")?
+        } else {
+            let attr_place = this.deref_operand(attr_op)?;
+            let attr_kind_place = attr_place.offset(Size::ZERO, MemPlaceMeta::None, i32_layout, this)?;
+            this.read_scalar(attr_kind_place.into())?.not_undef()?
+        };
+
+        let u32_layout = this.layout_of(this.tcx.types.u32)?;
+        let locked_count_place = mutex_place.offset(Size::from_bytes(4), MemPlaceMeta::None, u32_layout, this)?;
+        this.write_scalar(Scalar::from_u32(0), locked_count_place.into())?;
+
+        let mutex_kind_place = mutex_place.offset(Size::from_bytes(12), MemPlaceMeta::None, i32_layout, &*this.tcx)?;
+        this.write_scalar(kind, mutex_kind_place.into())?;
+
+        Ok(0)
+    }
+
+    fn pthread_mutex_lock(&mut self, mutex_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        check_ptr_target_min_size(this, mutex_op, 16)?;
+
+        let mutex = this.read_scalar(mutex_op)?.not_undef()?;
+        if this.is_null(mutex)? {
+            return this.eval_libc_i32("EINVAL");
+        }
+        let mutex_place = this.deref_operand(mutex_op)?;
+
+        let i32_layout = this.layout_of(this.tcx.types.i32)?;
+        let kind_place = mutex_place.offset(Size::from_bytes(12), MemPlaceMeta::None, i32_layout, this)?;
+        let kind = this.read_scalar(kind_place.into())?.not_undef()?;
+
+        let u32_layout = this.layout_of(this.tcx.types.u32)?;
+        let locked_count_place = mutex_place.offset(Size::from_bytes(4), MemPlaceMeta::None, u32_layout, this)?;
+        let locked_count = this.read_scalar(locked_count_place.into())?.to_u32()?;
+
+        if kind == this.eval_libc("PTHREAD_MUTEX_NORMAL")? {
+            if locked_count == 0 {
+                this.write_scalar(Scalar::from_u32(1), locked_count_place.into())?;
+                Ok(0)
+            } else {
+                throw_unsup_format!("Deadlock due to locking a PTHREAD_MUTEX_NORMAL mutex twice");
```

Comment:
It is also an interesting question which error we want to raise here. "unsupported" doesn't fit, this is an issue with the program going wrong, not something Miri does not support. "ub" is also wrong though, deadlocks are not UB.

I think the best option is to use a custom program error. Maybe we can equip [`Abort`](https://github.com/rust-lang/miri/blob/676c4f8c8cafe06c10263131da48c496717b6b8e/src/eval.rs#L56) with a message that indicates why execution got aborted?

Reply:
I think there should be a new custom exit for deadlocks. It's a different kind of error. Technically not even fatal, so it could just lint and continue

Reply:
If reasonably doable, a test for it would be great, too

Reply:
The only thing it can continue with is an infinite loop. We must not return from the current function as that would mean the lock was acquired twice. I don't see any use in that.^^ That's why an `Abort` felt appropriate. But I am also fine adding `Deadlock` as an extra variant to `TerminationInfo`.